sycl/test/matrix/matrix-bf16-test-SG-16.cpp

```diff
@@ -0,0 +1,188 @@
+#include <CL/sycl.hpp>
+#if (SYCL_EXT_ONEAPI_MATRIX == 2)
+#include <iostream>
+
+using namespace sycl;
+using namespace sycl::ext::oneapi::experimental::matrix;
+
+#define TILE_SZ 16
+#define TM (TILE_SZ-1)
+#define TN (TILE_SZ-1)
+#define TK (2*TILE_SZ-2)
+
+#define SG_SZ 16
+
+template <typename T, size_t NUM_ROWS, size_t NUM_COLS> struct big_matrix{
+public:
+  T *mat;
+
+public:
+  T *get_data() { return mat; }
+  void set_data(T *data) { mat = data; }
+  big_matrix(T *data) : mat(data) {
+  }
+};
+
+template <typename T1, typename T2, size_t NUM_ROWS_A, size_t NUM_COLS_A, size_t NUM_ROWS_B,
+          size_t NUM_COLS_B, size_t NUM_ROWS_C, size_t NUM_COLS_C>
+void matrix_multiply(big_matrix<T1, NUM_ROWS_C, NUM_COLS_C> &C, big_matrix<T2, NUM_ROWS_A, NUM_COLS_A> &A, big_matrix<T2, NUM_ROWS_B, NUM_COLS_B> &B) {
+  size_t M = NUM_ROWS_C;
+  size_t N = NUM_COLS_C;
+  size_t K = NUM_COLS_A;
+  // B => K/4 x N*4, A => M x K, C => M, N
+  // stride should be X's cols, e.g., B's stirde = N*4
+  assert(NUM_ROWS_C == NUM_ROWS_A && NUM_COLS_A == NUM_ROWS_B * 2);
+  size_t NDRangeM = M / TM;
+  size_t NDRangeN = N / TN;
+  buffer<unsigned short, 2> bufA(A.get_data(), range<2>(M, K));
+  buffer<unsigned short, 2> bufB(B.get_data(), range<2>(K, N));
+  buffer<float, 2> bufC((float*)C.get_data(), range<2>(M, N));
+
+  queue q;
+  q.submit([&](handler &cgh) {
+     auto accC = bufC.get_access<access::mode::read_write>(cgh);
+     auto accA = bufA.get_access<access::mode::read_write>(cgh);
+     auto accB = bufB.get_access<access::mode::read_write>(cgh);
+
+     cgh.parallel_for<class imatrix>(
+         nd_range<2>({NDRangeM, NDRangeN * SG_SZ}, {1, 1 * SG_SZ}),
+         [accA, accB, accC, M, N, K](nd_item<2> spmd_item) [[intel::reqd_sub_group_size(SG_SZ)]]
+
+         {
+           // The submatrix API has to be accessed by all the workitems in a
+           // subgroup these functions will be called once by the subgroup no
+           // code divergence between the workitems
+           const auto global_idx = spmd_item.get_global_id(0);
+           const auto global_idy = spmd_item.get_global_id(1);
+           const auto sg_startx = global_idx - spmd_item.get_local_id(0);
+           const auto sg_starty = global_idy - spmd_item.get_local_id(1);
+
+           ext::oneapi::sub_group sg = spmd_item.get_sub_group();
+           joint_matrix<ext::oneapi::sub_group, unsigned short, TM, TK> sub_a(sg);
+           // For B, since current implementation does not support non-packed layout,
+           // users need to specify the updated VNNI sizes along with the packed_b layout.
+           // By default, the layout is row_major and size is (TK, TN).
+           joint_matrix<ext::oneapi::sub_group, unsigned short, TK, TN, matrix_layout::packed_b> sub_b(sg);
+           joint_matrix<ext::oneapi::sub_group, float, TM, TN> sub_c(sg);
+
+           // Only the leader perform AMX computation.
+           if (spmd_item.get_local_id(1) % TILE_SZ)
+             return;
```

Comment:
Again, we should not have "the leader" test in the code. this is wrong because all work items in the subgroup must enter the joint matrix code. So we should not have diverge code in there. 
The OpenCL backend will not vectorize the code so we should be able to generate ONE AMX code for the whole sub group